From d6d4f27de2035490d18a76a6f23a9b370dc19c76 Mon Sep 17 00:00:00 2001
From: Stella Laurenzo <stellaraccident@gmail.com>
Date: Wed, 22 Jan 2025 13:28:47 -0800
Subject: [PATCH] Disable HIP_PLATFORM auto-detect if already defined.

This code appears to be old copy-pasta and is effectively just a no-op if -DHIP_PLATFORM=amd is passed explicitly. It has a number of nit-picky checks which will fail in this case for no real reason.

This is actually worse than that because it leaks global variables that establish a different default value/scoping vs the definitions in the inner hipamd/CMakeLists.txt. I opted to rewrite this section to scope the detection logic and keep it from leaking with respect to the inner build file. The detection logic is still not great but at least will not actively cause damage or make control flow hard to reason about. It seems like this is for legacy/compatibility so I opted to simply sequester it vs applying more diligence to it.
---
 CMakeLists.txt        | 43 +++++++++++++++++++++++++++----------------
 hipamd/CMakeLists.txt |  3 ++-
 2 files changed, 29 insertions(+), 17 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index cc9cf1654..ad2201073 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -40,40 +40,51 @@ endif()
 #############
 # Build steps
 #############
-if(CLR_BUILD_HIP)
-  if (UNIX)
-    set(HIPCC_EXECUTABLE "hipcc")
-    set(HIPCONFIG_EXECUTABLE "hipconfig")
-  else()
-    set(HIPCC_EXECUTABLE "hipcc.exe")
-    set(HIPCONFIG_EXECUTABLE "hipconfig.exe")
-  endif()
+
+# Attempt to auto-detect HIP_PLATFORM by interrogating hipconfig. This is kept
+# for compatibility: users are advised to pass HIP_PLATFORM explicitly.
+# Sets the HIP_PLATFORM variable in the parent scope.
+function(_hip_clr_auto_detect_hip_platform)
+    if (UNIX)
+      set(HIPCC_EXECUTABLE "hipcc")
+      set(HIPCONFIG_EXECUTABLE "hipconfig")
+    else()
+      set(HIPCC_EXECUTABLE "hipcc.exe")
+      set(HIPCONFIG_EXECUTABLE "hipconfig.exe")
+    endif()

     # Set default HIPCC_BIN_DIR to /opt/rocm/bin
     if(NOT DEFINED HIPCC_BIN_DIR AND UNIX)
-        set(HIPCC_BIN_DIR "/opt/rocm/bin" CACHE STRING "Default hipcc directory on linux.")
+        set(HIPCC_BIN_DIR "/opt/rocm/bin")
     endif()
-    message(STATUS "HIPCC Binary Directory: ${HIPCC_BIN_DIR}")
+    message(STATUS "Auto-detect HIP_PLATFORM from HIPCC Binary Directory: ${HIPCC_BIN_DIR}")

     if(NOT EXISTS ${HIPCC_BIN_DIR}/${HIPCONFIG_EXECUTABLE})
         message(FATAL_ERROR "Please pass hipcc/build or hipcc/bin using -DHIPCC_BIN_DIR.")
     endif()

-    message(STATUS "HIP Common Directory: ${HIP_COMMON_DIR}")
-    if(NOT DEFINED HIP_COMMON_DIR)
-        message(FATAL_ERROR "Please pass HIP using -DHIP_COMMON_DIR. HIP_COMMON_DIR is incorrect")
-    endif()
     # Determine HIP_PLATFORM
     if(NOT DEFINED HIP_PLATFORM)
         if(NOT DEFINED ENV{HIP_PLATFORM})
             execute_process(COMMAND ${HIPCC_BIN_DIR}/${HIPCONFIG_EXECUTABLE} --platform
-                OUTPUT_VARIABLE HIP_PLATFORM
+                OUTPUT_VARIABLE _detected_hip_platform
                 OUTPUT_STRIP_TRAILING_WHITESPACE)
+            set(HIP_PLATFORM "${_detected_hip_platform}" PARENT_SCOPE)
         else()
-            set(HIP_PLATFORM $ENV{HIP_PLATFORM} CACHE STRING "HIP Platform")
+            set(HIP_PLATFORM "$ENV{HIP_PLATFORM}" PARENT_SCOPE)
         endif()
     endif()
+endfunction()
+if(CLR_BUILD_HIP)
+    message(STATUS "HIP Common Directory: ${HIP_COMMON_DIR}")
+    if(NOT DEFINED HIP_COMMON_DIR)
+        message(FATAL_ERROR "Please pass HIP using -DHIP_COMMON_DIR. HIP_COMMON_DIR is incorrect")
+    endif()
+    if(NOT DEFINED HIP_PLATFORM)
+        _hip_clr_auto_detect_hip_platform()
+    endif()
 endif()
+
 if(CLR_BUILD_HIP)
     option(BUILD_SHARED_LIBS "Build the shared library" ON)
     if (NOT BUILD_SHARED_LIBS)
diff --git a/hipamd/CMakeLists.txt b/hipamd/CMakeLists.txt
index 9538dbbad..18cbb2a70 100755
--- a/hipamd/CMakeLists.txt
+++ b/hipamd/CMakeLists.txt
@@ -53,7 +53,8 @@ if(MSVC)
   set(CMAKE_SHARED_LINKER_FLAGS "${CMAKE_SHARED_LINKER_FLAGS} /DEBUG:FULL")
 endif()

-set(HIPCC_BIN_DIR "" CACHE STRING "HIPCC and HIPCONFIG binary directories")
+# Set default HIPCC_BIN_DIR to /opt/rocm/bin
+set(HIPCC_BIN_DIR "/opt/rocm/bin" CACHE STRING "HIPCC and HIPCONFIG binary directories")

 if(__HIP_ENABLE_PCH)
   set(_pchStatus 1)
--
2.43.0
